app/services/module_progress.rb

```diff
@@ -27,6 +27,13 @@ def completed?
     all?(mod.module_items)
   end
 
+  # Completed date for module
+  # @return [DateTime]
+  def completed_at
```

Comment:
Because this method is identical to lines inside `CourseProgress#completed_modules` a good addition is to refactor to use this new method 
```ruby
module_progess(mod).completed_at
```

Reply:
Currently:

```ruby
  # TODO: use a dedicated event for module completion
  # TODO: continue refactor to leverage ModuleProgress methods in this class
  #
  # @return [Array<Array>] Tabular data of completed training module
  def completed_modules
    by_state(:completed).map do |mod|
      final_page = mod.module_items.last.name
      completed_at = training_module_events(mod.name).where_properties(id: final_page).first.time

      [mod, completed_at]
    end
  end
```